app/components/confirmation-modal/component.js

```diff
@@ -0,0 +1,46 @@
+import Ember from 'ember';
+import layout from './template';
+
+export default Ember.Component.extend({
+  confirmationModalService: Ember.inject.service('confirmation-modal'),
+  layout: layout,
+
+  model: Ember.computed.reads('modal.model'),
+
+  init() {
+    this._super.apply(this, arguments);
+    this.get('confirmationModalService').on('open', (modal) => {
+      this.set('modal', modal);
+      this.set('isOpen', true);
+    });
+  },
+
+  close() {
+    this.set('isOpen', false);
+    this.set('modal', null);
+  },
+
+  actions: {
+
+    cancel() {
+      let modal = this.get('modal');
+      let result = modal.onCancel && modal.onCancel();
+      if (result && result.then) {
+        result.then(() => this.close());
+      } else {
+        this.close();
+      }
```

Comment:
Would it be simple to promisify the result instead of detecting that it is a thennable?
```
return Ember.RSVP.resolve().then(() => {
  return result;
}).then(() => {
  this.close();
});
```